include/deal.II/lac/relaxation_block.templates.h

```diff
@@ -151,6 +153,39 @@ RelaxationBlock<MatrixType,inverse_type>::invert_diagblocks ()
   this->inverses_computed(true);
 }
 
+namespace internal
+{
+  /**
+   * default implementation for serial vectors. Here we don't need to make a
+   * copy into a ghosted vector, so just return a reference to @p prev.
+   */
+  template <class VECTOR, class VEC1>
+  const VECTOR &
+  prepare_ghost_vector(
+    const VECTOR &prev,
+    const VEC1 * /*other*/)
+  {
+    return prev;
+  }
+
+  /**
+   * Specialization for Trilinos. Use the ghosted vector.
+   */
+  template <>
+  const TrilinosWrappers::MPI::Vector &
+  prepare_ghost_vector(
+    const TrilinosWrappers::MPI::Vector &prev,
+    const TrilinosWrappers::MPI::Vector *other)
+  {
+    Assert(other!=NULL,
+           ExcMessage("You need to provide a ghosted vector in RelaxationBlock::AdditionalData::temp_trilinos_ghost_vector."));
+    Assert(other->size()==prev.size(), ExcInternalError());
+
+    // import ghost values:
+    *const_cast<TrilinosWrappers::MPI::Vector *>(other) = prev;
+    return *other;
```

Comment:
I don't think we have a policy. The issues are actually separate. You want the `mutable` because you have a scratch variable that is used in `const` functions. You can avoid the `const_cast` by making the variable `mutable`. I think that also makes the declaration of the variable easier to read as it indicates its purpose.

As for the thread-safety: these issues are so notoriously difficult to track down that I'm usually tempted to just make things thread-safe if possible. Using a mutex around a significant is also not overly expensive as long as there are no other threads vying for it.